include/wil/coroutine.h

```diff
@@ -0,0 +1,791 @@
+    * co_await wil::com_task(GetNameAsync()); // now defaults to resume_same_apartment();
+    *
+    * You can store the task in a variable, but since it is a move-only
+    * object, you will have to use std::move in order to transfer ownership out of
+    * an lvalue.
+    *
+    * winrt::IAsyncAction SomethingAsync()
+    * {
+    *     wil::com_task<int> task;
+    *     switch (source)
+    *     {
+    *     // Some of these might return wil::task<int>,
+    *     // but assigning to a wil::com_task<int> will make
+    *     // the task resume in the same COM apartment.
+    *     case widget: task = GetValueFromWidgetAsync(); break;
+    *     case gadget: task = GetValueFromGadgetAsync(); break;
+    *     case doodad: task = GetValueFromDoodadAsync(); break;
+    *     default:     FAIL_FAST(); // unknown source
+    *     }
+    *     auto value = co_await std::move(task); // **** need std::move
+    *     DoSomethingWith(value);
+    * }
+    *
+    * You can wait synchronously by calling get(). The usual caveats
+    * about synchronous waits on STA threads apply.
+    *
+    * auto value = GetValueFromWidgetAsync().get();
+    * 
+    * auto task = GetValueFromWidgetAsync();
+    * auto value = std::move(task).get(); // **** need std::move
+    */
+
+// Detect which version of the coroutine standard we have.
+#if defined(_RESUMABLE_FUNCTIONS_SUPPORTED)
+#include <experimental/coroutine>
+#define __WI_COROUTINE_NAMESPACE ::std::experimental
+#elif defined(__cpp_lib_coroutine)
+#include <coroutine>
+#define __WI_COROUTINE_NAMESPACE ::std
+#else
+#error You must compile with C++20 coroutine support to use coroutine.h.
+#endif
+#include <atomic>
+#include <exception>
+#include <wil/wistd_memory.h>
+#include <wil/wistd_type_traits.h>
+#include <wil/result_macros.h>
+
+namespace wil
+{
+    // There are three general categories of T that you can
+    // use with a task. We give them these names:
+    //
+    // T = void ("void category")
+    // T = some kind of reference ("reference category")
+    // T = non-void non-reference ("object category")
+    //
+    // Take care that the implementation supports all three categories.
+    //
+    // There is a sub-category of object category for move-only types.
+    // We designed our task to be co_awaitable only once, so that
+    // it can contain a move-only type. Any transfer of T as an
+    // object category must be done as an rvalue reference.
+    template<typename T>
+    struct task;
+
+    template<typename T>
+    struct com_task;
+}
+
+namespace wil::details::coro
+{
+    template<typename T>
+    struct task_promise;
+
+    // Unions may not contain references, C++/CX types, or void.
+    // To work around that, we put everything inside a result_wrapper
+    // struct, and put the struct in the union. For void,
+    // we create a special empty structure.
+    //
+    // get_value returns rvalue reference to T for object
+    // category, or just T itself for void and reference

```

Comment:
"get_value returns rvalue reference to T for object category" is incorrect (stale comment?)

In fact, `get_value()` is clearly destructive for move-only objects. Would it be reasonable to make it an rvalue method, along with `client_await_resume()`?